pkg/registry/rbac/validation/kcp.go

```diff
@@ -0,0 +1,83 @@
+package validation
+
+import (
+	"context"
+	"strings"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	authserviceaccount "k8s.io/apiserver/pkg/authentication/serviceaccount"
+	"k8s.io/apiserver/pkg/authentication/user"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+const (
+	// ScopeExtraKey is the key used in a user's "extra" to specify
+	// that the user is restricted to a given scope. Valid values are:
+	// - "cluster:<name>"
+	// In the future, we might add:
+	// - "interval:from:to".
+	// Scopes are and'ed. Scoping to multiple clusters invalidates it for all.
+	ScopeExtraKey = "authentication.kcp.io/scopes"
+)
+
+type appliesToUserFunc func(user user.Info, subject rbacv1.Subject, namespace string) bool
+type appliesToUserFuncCtx func(ctx context.Context, user user.Info, subject rbacv1.Subject, namespace string) bool
+
+var appliesToUserWithWarrants = withScopes(appliesToUser)
+
+// withScopes wraps the appliesToUser predicate to check for the base user and any warrants.
+func withScopes(appliesToUser appliesToUserFunc) appliesToUserFuncCtx {
+	var recursive appliesToUserFuncCtx
+	recursive = func(ctx context.Context, u user.Info, bindingSubject rbacv1.Subject, namespace string) bool {
+		var clusterName logicalcluster.Name
+		if cluster := genericapirequest.ClusterFrom(ctx); cluster != nil {
+			clusterName = cluster.Name
+		}
+		if IsInScope(u, clusterName) && appliesToUser(u, bindingSubject, namespace) {
+			return true
+		}
+
+		return false
+	}
+	return recursive
+}
+
+// IsServiceAccount returns true if the user is a service account.
+func IsServiceAccount(attr user.Info) bool {
+	return strings.HasPrefix(attr.GetName(), "system:serviceaccount:")
+}
+
+// IsForeign returns true if the service account is not from the given cluster.
+func IsForeign(attr user.Info, cluster logicalcluster.Name) bool {
+	clusters := attr.GetExtra()[authserviceaccount.ClusterNameKey]
+	if clusters == nil {
+		// an unqualified service account is considered local: think of some
+		// local SubjectAccessReview specifying a service account without the
+		// cluster scope.
+		return false
+	}
+	return !sets.New(clusters...).Has(string(cluster))
+}
+
+// IsInScope checks if the user is valid for the given cluster.
+func IsInScope(attr user.Info, cluster logicalcluster.Name) bool {
+	if IsServiceAccount(attr) && IsForeign(attr, cluster) {
+		return false
+	}
+
+	scopes := attr.GetExtra()[ScopeExtraKey]
+	for _, scope := range scopes {
+		switch {
+		case strings.HasPrefix(scope, "cluster:"):
+			if scope != "cluster:"+string(cluster) {
+				return false
+			}
+		default:
+			// Unknown scope, ignore.
+		}
+	}
```

Comment:
Now I think this is wrong. If I have 2 scopes to 2 separate clusters (Im still in scope for both), but as soon as I get evaluated to one of them which is not the one Im targeting now - I get false and deny.

Reply:
```
scopes:
   cluster:foo
   cluster:bar
```
`InScope(user, `bar`)  and I get `false` in first check as `foo != bar`

Reply:
that's intentional. You have a scoped user, and now you evaluate it inside another cluster. Then this user is equivalent to `{Name: "system:anonymous", Groups: []string{"system:authenticated"}}`.